sys/sema/sema.c

```diff
@@ -97,6 +103,63 @@ int _sema_wait(sema_t *sema, int block, uint64_t us)
 
     return 0;
 }
+#endif
+
+#if IS_USED(MODULE_ZTIMER)
+int _sema_wait_ztimer(sema_t *sema, int block,
+                      ztimer_clock_t *clock, uint32_t timeout)
+{
+    assert(sema != NULL);
+
+    if (sema->state != SEMA_OK) {
+        return -ECANCELED;
+    }
+
+    int did_block = block;
+    unsigned old = irq_disable();
+    while ((sema->value == 0) && block) {
+        irq_restore(old);
+        if (timeout == 0) {
+            mutex_lock(&sema->mutex);
+        }
+        else {
+            ztimer_now_t start = ztimer_now(ZTIMER_MSEC);
+            block = !ztimer_mutex_lock_timeout(ZTIMER_MSEC, &sema->mutex, timeout);
+            uint32_t elapsed = (uint32_t)(ztimer_now(ZTIMER_MSEC) - start);
```

Comment:
Can't simply this part be abstracted depending on the timer backend?

Reply:
Untested but how about something like:

```diff
diff --git a/sys/sema/sema.c b/sys/sema/sema.c
index 05f61f9380..42e7c9e133 100644
--- a/sys/sema/sema.c
+++ b/sys/sema/sema.c
@@ -29,6 +29,12 @@
 #define ENABLE_DEBUG 0
 #include "debug.h"
 
+#if IS_USED(MODULE_XTIMER)
+typedef uint64_t _sema_timeout_t;
+#elif IS_USED(MODULE_ZTIMER)
+typedef uint32_t _sema_timeout_t;
+#endif
+
 void sema_create(sema_t *sema, unsigned int value)
 {
     assert(sema != NULL);
@@ -49,67 +55,13 @@ void sema_destroy(sema_t *sema)
     mutex_unlock(&sema->mutex);
 }
 
-#if !IS_USED(MODULE_SEMA_NO_XTIMER)
-int _sema_wait(sema_t *sema, int block, uint64_t us)
+#if IS_USED(MODULE_XTIMER) || IS_USED(MODULE_ZTIMER)
+int _sema_wait(sema_t *sema, int block, void* clock, _sema_timeout_t timeout)
 {
     assert(sema != NULL);
-
-    if (sema->state != SEMA_OK) {
-        return -ECANCELED;
-    }
-
-    int did_block = block;
-    unsigned old = irq_disable();
-    while ((sema->value == 0) && block) {
-        irq_restore(old);
-        if (us == 0) {
-            mutex_lock(&sema->mutex);
-        }
-#if IS_USED(MODULE_XTIMER)
-        else {
-            uint64_t start = xtimer_now_usec64();
-            block = !xtimer_mutex_lock_timeout(&sema->mutex, us);
-            uint64_t elapsed = xtimer_now_usec64() - start;
-
-            if (elapsed < us) {
-                us -= elapsed;
-            }
-            else {
-                block = 0;
-            }
-        }
-#endif
-
-        if (sema->state != SEMA_OK) {
-            mutex_unlock(&sema->mutex);
-            return -ECANCELED;
-        }
-
-        old = irq_disable();
-    }
-
-    if (sema->value == 0) {
-        irq_restore(old);
-        return (did_block) ? -ETIMEDOUT : -EAGAIN;
-    }
-
-    unsigned int value = --sema->value;
-    irq_restore(old);
-
-    /* only unlock mutex if it was a blocking operation */
-    if (did_block && value > 0) {
-        mutex_unlock(&sema->mutex);
-    }
-
-    return 0;
-}
-#endif
-
 #if IS_USED(MODULE_ZTIMER)
-int _sema_wait_ztimer(sema_t *sema, int block,
-                      ztimer_clock_t *clock, uint32_t timeout)
-{
-    assert(sema != NULL);
+    assert(clock != NULL);
+#endif
 
     if (sema->state != SEMA_OK) {
         return -ECANCELED;
@@ -123,10 +75,15 @@ int _sema_wait_ztimer(sema_t *sema, int block,
             mutex_lock(&sema->mutex);
         }
         else {
+#if IS_USED(MODULE_XTIMER)
+            uint64_t start = xtimer_now_usec64();
+            block = !xtimer_mutex_lock_timeout(&sema->mutex, timeout);
+            uint64_t elapsed = xtimer_now_usec64() - start;
+#elif IS_USED(MODULE_ZTIMER)
             ztimer_now_t start = ztimer_now(ZTIMER_MSEC);
             block = !ztimer_mutex_lock_timeout(ZTIMER_MSEC, &sema->mutex, timeout);
             uint32_t elapsed = (uint32_t)(ztimer_now(ZTIMER_MSEC) - start);
-
+#endif
             if (elapsed < timeout) {
                 timeout -= elapsed;
             }
@@ -158,8 +115,6 @@ int _sema_wait_ztimer(sema_t *sema, int block,
 
     return 0;
 }
-#endif
-
 
 int sema_post(sema_t *sema)
 {
```

Reply:
The headers need quite some adaptation with though.

Reply:
Mainly wrap all `time` related functionality in `#if IS_USED(MODULE_XTIMER) ||  IS_USED(MODULE_XTIMER) || defined(DOXYGEN)`, and then add some `@pre ztimer or xtimer must be included` to those function definitions.

Reply:
Don't think this works well in cases where both ztimer and xtimer are used simultaneously. Imagine some library uses the `sema_wait_timed()` with an uint64_t timeout value, while some other library uses the `sema_wait_timed_ztimer()` function in the same build. As this should be possible, I think there is no way around keeping both API-paths intact for now.

Reply:
Yes you are right, I missed this answer before. Lets keep them separate then.